### ISSUE TITLE
Help with creating own program that needs to read iconcache

### ISSUE THREAD

**Issue by cegredev** [thumbcacheviewer/thumbcacheviewer#6]
**Disclaimer:** I know this is kind of an unusual request and that this is not Stack Overflow, so feel free to ignore it. ;-)

I am currently writing a small file manager as a hobby project and need to be able to read in the icon for any given path from iconcache_x.db. I am amazed by what you have created here as there seems to be no documentation on the format anywhere on the internet and would like to ask if you could help or even just point me in the right direction with that.

I have read the [previous issue on this topic](https://github.com/thumbcacheviewer/thumbcacheviewer/issues/1) but can't seem to get the mapping to work (not sure what you're expected to put into the initial scan directory field) and am also really unfamiliar with C++, so reading the code is quite tricky for me (currently writing the backend for the project in Rust).

In short what I need to do is this: Get the icons for all files in a directory in a batch, so for example:

1. The user clicks on the "Downloads" folder
2. All files in the downloads folder are retrieved
3. For each of those files, the icon is looked up and returned

I'd be really thankful if you could outline an algorithm or point out a snippet of code I should take a look at to get this behavior. If not, I'll just have to try and understand the algorithm in the cmd version, as reading a single file should be at least a bit easier than an entire code base and it seems to contain the feature I need.

No matter how you decide, thanks in advance!

**Comment by thumbcacheviewer** [comment]
I don't think you need to process the iconcache databases to accomplish the icon retrieval. There's a Windows API that'll return the icon for a file: [SHGetFileInfo](https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shgetfileinfow)

This example returns the small icon:
```
SHFILEINFO sfi;
SHGetFileInfo( "C:\\example.jpg", FILE_ATTRIBUTE_NORMAL, &sfi, sizeof( SHFILEINFO ), SHGFI_USEFILEATTRIBUTES | SHGFI_ICON | SHGFI_SMALLICON );
```

sfi.hIcon will be the handle for the icon. There's GDI functions that can turn it into a bitmap if that's what you need. I'm not sure how Rust works with the Win32 API, but this would be the best way to get the icon. The iconcache isn't guaranteed to have the icon of a file cached so it's not 100% reliable.

To be able to retrieve the icon from the iconcache, you would have to get the file's thumbcache ID. That algorithm for that is here: https://github.com/thumbcacheviewer/thumbcacheviewer/blob/10ce835b11155a9f1fc66aa258404c1229109993/thumbcache_viewer/map_entries.cpp#L184
You would then need to scan the database for that hash value and extract the image based on that entry's information. It's a lot more work than simply using the SHGetFileInfo API.

In either method you would still have to use Win32 APIs. SHGetFileInfo is probably the best approach.

I don't have any experience with Rust to provide code, but hopefully this info can help.

**Comment by cegredev** [comment]
Thanks for your reply!

> There's a Windows API that'll return the icon for a file: SHGetFileInfo

Interesting, I did not know about that! What I previously tried was getting the associated executable for a file type through the registry editor and then calling `ExtractAssociatedIcon` in PowerShell on that .exe to get its icon, but that was very fragile and didn't really work. I'm surprised I didn't stumble upon this method earlier.

> The iconcache isn't guaranteed to have the icon of a file cached so it's not 100% reliable.

Also did not know that, although it makes sense now that I think about it.

> SHGetFileInfo is probably the best approach.

Since it really seems to be a lot of work to go through the iconcache you're probably right, thanks for your help!

...I have just one more question: Where do I actually find the icon after calling the function, since the return value only seems to indicate success or failure?

**Comment by thumbcacheviewer** [comment]
One of the variables (the third parameter) that's passed into the SHGetFileInfo function is a [SHFILEINFO](https://docs.microsoft.com/en-us/windows/win32/api/shellapi/ns-shellapi-shfileinfoa) structure. It's filled out when the function successfully returns. In that structure there's an HICON variable that's a handle to the icon. If you use any GDI drawing functions in your program, you can draw it to the screen with [DrawIcon](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-drawicon).

**Comment by cegredev** [comment]
Oh, my bad, should have read the documentation more carefully. Thanks for your help!
